onnxruntime/core/framework/execution_providers.h

```diff
@@ -99,5 +105,9 @@ class ExecutionProviders {
 
   // maps for fast lookup of an index into exec_providers_
   std::unordered_map<std::string, size_t> provider_idx_map_;
+
+  // Whether the CPU provider was implicitly added to a session for fallback (true),
+  // or whether it was explicitly added by the caller.
+  bool cpu_execution_provider_was_implicitly_added_ = false;
```

Comment:
This seemed the least bad place to put this extra state. The InferenceSession knows this and needs to pass it to SessionState so Session Finalize can VerifyEachNodeIsAssignedToAnEp. I originally tried passing it as a bool flag to Session state, but that required updating way too many places in the test code which initialized a session directly.